### PR TITLE
Add ES proxy service

### PR DESCRIPTION
**Pull request by nsdavidson**
So long NAT gateways and thanks for all the fish!
- adds a standalone nginx service to each server to proxy & sign ElasticSearch requests
- also some good bug fixes and cleanups

### CODE REVIEW

File: files/aws_creds.lua [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    return {
+      aws_access_key_id = shd:get("aws_access_key_id"),
+      aws_secret_access_key = shd:get("aws_secret_access_key"),
+      aws_security_token = shd:get("aws_security_token")
+    }
+  end
+end
+
+function _M.get_iam_credentials_uncached()
+  -- You need /_awsapi proxied to http://169.254.169.254 for this to work
```
Comment (irvingpop) [inline files/aws_creds.lua:51]: Curious if this bit is a problem?

File: files/es_auth.lua [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+local cont_type = ngx.req.get_headers()["Content-Type"]
+if cont_type == nil then
+  cont_type = ""
+end
+
+local config = {
+  aws_host       = "elasticsearch",
+  aws_key        = creds['aws_access_key_id'],
+  aws_secret     = creds['aws_secret_access_key'],
+  aws_region     = "us-east-1",
```
Comment (irvingpop) [inline files/es_auth.lua]: should we template this out to support other AWS regions?

File: files/nginx-esproxy.conf [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  gzip_comp_level 2;
+  gzip_proxied any;
+  gzip_types text/plain text/css application/x-javascript text/xml application/xml application/xml+rss text/javascript application/json;
+
+  lua_package_path "/var/opt/opscode/nginx-esproxy/etc/scripts/?.lua;$prefix/?.lua;;";
+
+  client_body_temp_path "/var/opt/opscode/nginx-esproxy/tmp/client_body";
+
+  upstream elasticsearch {
+    server <%= @elasticsearch_server %>:443;
```
Comment (irvingpop) [inline files/nginx-esproxy.conf]: whoops ERB still here

File: files/aws_creds.lua (hunk 1 above)
Comment (irvingpop) [reply files/aws_creds.lua:51]: oh I see the _awsapi proxy below - necessary?

File: backendless_chef.yaml [hunk 4]
```diff
@@ -578,11 +578,9 @@ Resources:
         Version: "2012-10-17"
         Statement:
           - Effect: "Allow"
-            Principal: "*"
+            Principal:
+              AWS: !GetAtt ChefRole.Arn
```
Comment (irvingpop) [inline backendless_chef.yaml]: I'm wondering if this should be:
```
!If [CreateChefServerIamRole, !GettAtt ChefRole.Arn, !Ref ChefServerIamRole]
```

but now I'm wondering what you're even supposed to pass in as `ChefServerIamRole`  (is it an Arn? or just a role name?)

Comment (irvingpop) [reply backendless_chef.yaml]: the second one might have to be `!Sub arn:aws:iam::${AWS::AccountId}:role/${ChefServerIamRole}`

Comment (nsdavidson) [reply backendless_chef.yaml]: Yeah, according to the docs it needs the ARN.  Updated to account for a user-provided IAM role.

**Comment by irvingpop** [comment]
wondering if we should add a FilesLocation parameter for all the nginx/lua files plus move main.sh there so that it's like:

```
source: !Sub ${FilesLocation}nginx-esproxy.conf
```

That would allow us to cleanly switch between branches for testing, and also allow customers to stage their own files if they ever desired (no internet access case)

**Comment by nsdavidson** [comment]
Yeah, we should absolutely do that just for testing if nothing else.  

The no internet access case would currently still be broken by the openresty install because we add an external repo to pull that from.

**Pull request merged**